### ISSUE TITLE
Darinrikusham - when user deposits assets during lock period they will rely on vault to open to determine how many shares will be allocated to them as there is no method to determine shares during lock period and after the `open` function is called by protocol, users will come to know shares allocated to them and after that if users are not convinced with it and want to redeem those shares and get deposited asset back they will receive assets less then deposited due to calculation method implimented in `previewSettle` function for `sharesToMint` calculation.

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2024-03-amphor-judging#112]
Darinrikusham

high

# when user deposits assets during lock period they will rely on vault to open to determine how many shares will be allocated to them as there is no method to determine shares during lock period and after the `open` function is called by protocol, users will come to know shares allocated to them and after that if users are not convinced with it and want to redeem those shares and get deposited asset back they will receive assets less then deposited due to calculation method implimented in `previewSettle` function for `sharesToMint` calculation.

## Summary
Users deposit funds using `requestDeposit` during lock period and they receive shares when `open` is called by protocol as there is no method to determine how many share will be received for amount users want to deposit in `lock` period. After that if users are not convinced with share allocation and calls `redeem` for those shares they do not get the same amount of assets as deposited back due to calculation method implimented in [previewSettle](https://github.com/sherlock-audit/2024-03-amphor/blob/6c797025ffe296e04607abf74400ff2bb36a7de3/asynchronous-vault/src/AsyncSynthVault.sol#L626) function, which they should be able to get as they don't know beforehand how many shares they will get.

## Vulnerability Detail
The number of shares to be minted for the amount of assets deposited during lock period is determined by following calculation when `open` is called on vault by protocol [https://github.com/sherlock-audit/2024-03-amphor/blob/6c797025ffe296e04607abf74400ff2bb36a7de3/asynchronous-vault/src/AsyncSynthVault.sol#L649](https://github.com/sherlock-audit/2024-03-amphor/blob/6c797025ffe296e04607abf74400ff2bb36a7de3/asynchronous-vault/src/AsyncSynthVault.sol#L649)

Users will know how much shares they have been allocated only after this `open` call and shares are minted and they claim it, as there is no function to determine the number of shares they will receive during lock period.

If after this in open state users are not convinced with the shares they have been allocated for their deposit amount and they want to redeem those shares they have been allocated and have their deposited assets back they will call `redeem` function on [`redeem`](https://github.com/sherlock-audit/2024-03-amphor/blob/6c797025ffe296e04607abf74400ff2bb36a7de3/asynchronous-vault/src/SyncSynthVault.sol#L359) line and doing the calculation on [`this-line`](https://github.com/sherlock-audit/2024-03-amphor/blob/6c797025ffe296e04607abf74400ff2bb36a7de3/asynchronous-vault/src/SyncSynthVault.sol#L603) they recieve assets amount which is less then what they have deposited.

Example scenario -

- Protocol deposits `1 weth` to the vault and they will be allocated `1 Share` for it.
- User-A deposits `3 weth` to the vault and doing calculation User-A will be allocated `3 Shares` for it.
- Vault value `4weth : 4 shares`
- Protocol calls `close` and the vault is locked.
- User-B calls [`requestDeposit`](https://github.com/sherlock-audit/2024-03-amphor/blob/6c797025ffe296e04607abf74400ff2bb36a7de3/asynchronous-vault/src/AsyncSynthVault.sol#L439C14-L439C28) for `11 weth` during this period and will be waiting for protocol to `open` it to determine how many shares User-B will receive.
- User-A calls [`requestRedeem`](https://github.com/sherlock-audit/2024-03-amphor/blob/6c797025ffe296e04607abf74400ff2bb36a7de3/asynchronous-vault/src/AsyncSynthVault.sol#L477C14-L477C27) for `2 Shares` and will be waiting for protocol to `open` in order to receive assets back.
- Protocol calls [`open`](https://github.com/sherlock-audit/2024-03-amphor/blob/6c797025ffe296e04607abf74400ff2bb36a7de3/asynchronous-vault/src/AsyncSynthVault.sol#L307) as for example assuming vault is in profit and `assetReturned` parameter is passed as `7 weth` by protocol and calculation is done by [`previewSettle`](https://github.com/sherlock-audit/2024-03-amphor/blob/6c797025ffe296e04607abf74400ff2bb36a7de3/asynchronous-vault/src/AsyncSynthVault.sol#L626) for amount of pending deposits and withdraw and by doing calculation for [`sharesToMint`](https://github.com/sherlock-audit/2024-03-amphor/blob/6c797025ffe296e04607abf74400ff2bb36a7de3/asynchronous-vault/src/AsyncSynthVault.sol#L649) and [`assetsToWithdraw`](https://github.com/sherlock-audit/2024-03-amphor/blob/6c797025ffe296e04607abf74400ff2bb36a7de3/asynchronous-vault/src/AsyncSynthVault.sol#L656) `6 shares` will be minted for User-B for `11 weth` deposited and `3 weth` will be allocated for User-A for for `2 Shares` deposited as per calculation as `fees` would `round-down` to 0.
- Vault value `15 weth and 8 shares`
- After this user-B will claim `6 shares` from `claimable-silo` and if User-B is not convinced with the share allocation and wants to `redeem` his shares and by doing [calculation](https://github.com/sherlock-audit/2024-03-amphor/blob/6c797025ffe296e04607abf74400ff2bb36a7de3/asynchronous-vault/src/SyncSynthVault.sol#L603) User-B will receive `10 weth` for redeeming `6 shares` which is less then `11 weth` which was deposited.

## Impact
Loss of assets for depositors during lock state of vault, as there is no method for them to calculate the shares they will receive for the amount of assets they deposit during `lock` period and can only be determined after vault is in open state, after calling `open` for the vault by the protocol if they are not convinced by share allocation and wants to redeem those shares to get the assets deposited back they will receive less amount then deposited back due to share calculation method in `previewSettle` function.

## Code Snippet
[https://github.com/sherlock-audit/2024-03-amphor/blob/6c797025ffe296e04607abf74400ff2bb36a7de3/asynchronous-vault/src/AsyncSynthVault.sol#L649-L651](https://github.com/sherlock-audit/2024-03-amphor/blob/6c797025ffe296e04607abf74400ff2bb36a7de3/asynchronous-vault/src/AsyncSynthVault.sol#L649-L651)

## Tool used

Manual Review

## Recommendation
A potential fix is to use `Math.Rounding.Ceil` instead of `Math.Rounding.Floor` in [`previewSettle`](https://github.com/sherlock-audit/2024-03-amphor/blob/6c797025ffe296e04607abf74400ff2bb36a7de3/asynchronous-vault/src/AsyncSynthVault.sol#L626) function at calculation for [sharesToMint ](https://github.com/sherlock-audit/2024-03-amphor/blob/6c797025ffe296e04607abf74400ff2bb36a7de3/asynchronous-vault/src/AsyncSynthVault.sol#L650) or protocol should implement different functions through which users can determine what will be the value in shares for the amount of asset they are requesting to deposit as it is not possible to determine currently during lock period of vault.

Duplicate of #131
